x-pack/plugins/endpoint/server/endpoint_app_context_services.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { IndexPatternRetriever } from './index_pattern';
+import { AgentService } from '../../ingest_manager/common/types';
+
+/**
+ * A singleton that holds shared services that are initialized during the start up phase
+ * of the plugin lifecycle. And stop during the stop phase, if needed.
+ */
+class EndpointAppContextService {
+  private indexPatternRetriever: IndexPatternRetriever | undefined;
+  private agentService: AgentService | undefined;
+
+  public start(dependencies: {
+    indexPatternRetriever: IndexPatternRetriever;
+    agentService: AgentService;
+  }) {
+    this.indexPatternRetriever = dependencies.indexPatternRetriever;
+    this.agentService = dependencies.agentService;
+  }
+
+  public stop() {}
+
+  public getAgentService(): AgentService {
+    if (!this.agentService) {
+      throw new Error(`must call start on ${EndpointAppContextService.name} to call getter`);
```

Comment:
If you must call `start` before calling `getAgentService`, why not just move the contents of `start` to the constructor? This will avoid sequential coupling.

Reply:
> If you must call start before calling getAgentService, why not just move the contents of start to the constructor? This will avoid sequential coupling.

It is stated in the class documentation that is a singleton and it state is global and that the object it holds are tied to the lifecycle start and stop phases. This is how it was in other plugins. If you have another way to have this single in a way that we can also mock out the services in it during testing I am willing to try it out.